File: ce/5.py
import sys

with open(sys.argv[1]) as f:
  for line in f.read().strip().splitlines():
    nn = tuple(map(int, line.split()))
    d = {}
    for k, v in zip(nn[:-1], nn[1:]):
      if d.get(k, None) == v:
        x = k
        sys.stdout.write(str(x))
        x = d[x]
        while x != k:
          sys.stdout.write(" " + str(x))
          x = d[x]
        sys.stdout.write("\n")
        break
      d[k] = v

File: ce/5.rb
File.open(ARGV[0], "r") do |f|
  while line = f.gets
    d = {}
    ii = line.split.map(&:to_i).to_a
    ii[0..-2].zip(ii[1..-1]).each do |(key, value)|
      if d[key] == value
        x = key
        print "#{x}"
        x = value
        while x != key
          print " #{x}"
          x = d[x]
        end
        puts
        break
      end
      d[key] = value
    end
  end
end
